src/bootstrap/src/core/builder.rs

```diff
@@ -473,14 +459,39 @@ impl StepDescription {
             return;
         }
 
-        // Handle all PathSets.
-        for (desc, should_run) in v.iter().zip(&should_runs) {
-            let pathsets = should_run.pathset_for_paths_removing_matches(&mut paths, desc.kind);
-            if !pathsets.is_empty() {
-                desc.maybe_run(builder, pathsets);
+        // Special type that helps bootstrap to:
+        //     1. keep the given CLI order on step execution
+        //     2. keep all matching crates so we can pass them to `Step::make_run` at once,
+        //        rather than calling it many times with a single crate.
+        let mut steps_to_run: HashMap<PathBuf, (&StepDescription, Vec<PathSet>)> = HashMap::new();
+
+        // Handle all PathSets
+        for path in &paths {
+            for (desc, should_run) in v.iter().zip(&should_runs) {
+                let pathsets = should_run.resolve_pathsets(path, desc.kind);
+                if !pathsets.is_empty() {
+                    steps_to_run
+                        .entry(path.clone())
+                        .or_insert_with(|| (desc, Vec::new()))
+                        .1
+                        .extend(pathsets);
+                }
             }
         }
 
+        // Start given steps and remove them from `paths`.
+        paths.retain(|path| {
+            if let Some((desc, pathsets)) = steps_to_run.remove(path) {
+                desc.maybe_run(builder, pathsets);
+
+                false
+            } else {
+                true
+            }
+        });
+
+        assert!(steps_to_run.is_empty(), "Some steps did not executed.");
```

Comment:
Nit: "some steps were not executed".
Discussion: In case of failure, should it also print which steps were not executed?